ext/storage_sources/s3_store/s3_store.cpp

```diff
@@ -0,0 +1,253 @@
+                } else {
+                    std::cout << "No objects in bucket." << std::endl;
+                }
+                std::cout << std::endl;
+            }
+
+            /* Delete object. */
+            fs->conn->delete_object(first_bucket, "WiredTiger.turtle");
+
+            /* List objects again. */
+            if (fs->conn->list_objects(first_bucket, bucket_objects)) {
+                std::cout << "Objects in bucket '" << first_bucket << "':" << std::endl;
+                if (bucket_objects.size() >= 1) {
+                    for (Aws::S3Crt::Model::Object &object : bucket_objects) {
+                        std::cout << "  * " << object.GetKey() << std::endl;
+                    }
+                } else {
+                    std::cout << "No objects in bucket." << std::endl;
+                }
+                std::cout << std::endl;
+            }
+        } else {
+            std::cout << "No buckets in AWS account." << std::endl;
+        }
+    }
+
+    *file_systemp = &fs->file_system;
+   return 0;
+}
+
+/*
+ * s3_fs_terminate --
+ *      Discard any resources on termination of the file system.
+ */
+static int
+s3_fs_terminate(WT_FILE_SYSTEM *file_system, WT_SESSION *session)
+{
+    S3_FILE_SYSTEM *s3_fs;
+
+    UNUSED(session); /* unused */
+
+    s3_fs = (S3_FILE_SYSTEM *)file_system;
+    delete(s3_fs->conn);
+    free(s3_fs);
+
+    return (0);
+}
+
+/*
+ * s3_add_reference --
+ *     Add a reference to the storage source so we can reference count to know when to really
+ *     terminate.
+ */
+static int
+s3_add_reference(WT_STORAGE_SOURCE *storage_source)
+{
+    UNUSED(storage_source);
+    return (0);
+}
+
+/*
+ * s3_terminate --
+ *     Discard any resources on termination.
+ */
+static int
+s3_terminate(WT_STORAGE_SOURCE *storage, WT_SESSION *session)
+{
+    /* NEED A WAY to pass the options to the ShutdownAPI call */
+    S3_STORAGE *s3;
+    s3 = (S3_STORAGE *)storage;
+
+    Aws::ShutdownAPI(options);
+
+    free (s3);
+    return (0);
+}
+
+/*
+ * wiredtiger_extension_init --
+ *     A S3 storage source library.
+ */
+int
+wiredtiger_extension_init(WT_CONNECTION *connection, WT_CONFIG_ARG *config)
+{
+    S3_STORAGE *s3;
+    int ret;
+
+    if ((s3 = (S3_STORAGE*)calloc(1, sizeof(S3_STORAGE))) == NULL)
+        return (errno);
+
+    s3->wt_api = connection->get_extension_api(connection);
+    UNUSED(config);
+
+    Aws::InitAPI(options);
+
+    /*
+     * Allocate a s3 storage structure, with a WT_STORAGE structure as the first field, allowing
+     * us to treat references to either type of structure as a reference to the other type.
+     */
+    s3->storage_source.ss_customize_file_system = s3_customize_file_system;
+    s3->storage_source.ss_add_reference = s3_add_reference;
+    s3->storage_source.terminate = s3_terminate;
+
+    /* Load the storage */
+    if ((ret = connection->add_storage_source(connection, "s3_store", &s3->storage_source, NULL)) !=
+      0)
+        free(s3);
```

Comment:
Should we define a err tag and use WT_ERROR macro?

Reply:
We will add error handling and logging in a ticket in the next sprint.